SCAutolib/controller.py

```diff
@@ -0,0 +1,419 @@
+                    f"cert-{user_.username}.pem")
+
+            if isinstance(user_, user.IPAUser):
+                self.ipa_ca.request_cert(csr, user_.username, user_.cert)
+            else:
+                self.local_ca.request_cert(csr, user_.username, user_.cert)
+
+        user_.card.enroll()
+
+        dump_to_json(user_.card)
+        # Update the dump for current user to link it with the card
+        dump_to_json(user_)
+
+    def cleanup(self):
+        if self.local_ca:
+            self.local_ca.cleanup()
+        if self.ipa_ca:
+            self.ipa_ca.cleanup()
+
+    def _validate_configuration(self, params: {} = None):
+        """
+        Validate schema of the configuration file. If some value doesn't present
+        in the config file, this value would be looked in the CLI parameters
+
+        :param params: CLI arguments
+        :return:
+        """
+        # FIXME: any schema requires all values to be in the config file, and
+        #  only IP address of IPA server is accepted from CLI arguments.
+        #  Add loading of the values from params dict
+        # IP regex
+        # Specify validation schema for CAs
+        schema_cas = Schema(And(
+            Use(dict),
+            # Check that CA section contains at least one and maximum
+            # two entries
+            lambda l: 1 <= len(l.keys()) <= 2,
+            {Optional("local_ca"): {
+                Optional("dir", default=Path("/etc/SCAutolib/ca")): Use(Path)},
+                Optional("ipa"): {
+                    'admin_passwd': Use(str),
+                    'root_passwd': Use(str),
+                    Optional('ip_addr', default=None): Use(str),
+                    'server_hostname': Use(str),
+                    'client_hostname': Use(str),
+                    'domain': Use(str),
+                    'realm': Use(str.upper)}}),
+            ignore_extra_keys=True)
+
+        # Specify validation schema for all users
+        schema_user = Schema({'name': Use(str),
+                              'passwd': Use(str),
+                              'pin': Use(str),
+                              Optional('card_dir', default=None): Use(Path),
+                              'card_type': Or("virtual", "real", "removinator"),
+                              Optional('cert', default=None): Use(Path),
+                              Optional('key', default=None): Use(Path),
+                              'local': Use(bool)})
+
+        # Specify general schema for whole config file
+        schema = Schema({"root_passwd": Use(str),
+                         "ca": schema_cas,
+                         "users": [schema_user]})
+
+        return schema.validate(self.lib_conf)
+
+    @staticmethod
+    def _general_steps_for_virtual_sc():
+        """
+        Prepare the system for virtual smart card. Preparation means to
+        configure pcscd service and opensc module to be able correctly working
```

Comment:
`be able to work correctly` or just `to work correctly`

Reply:
Referenced in issue #92

Reply:
Referenced in issue #92